src/Features/CSharp/Portable/Completion/CompletionProviders/SpeculativeTCompletionProvider.cs

```diff
@@ -76,19 +80,25 @@ private async Task<bool> ShouldShowSpeculativeTCompletionItemAsync(Document docu
                 // and the generic won't be "partially written".
                 if (testPosition == position)
                 {
-                    var typeArgumentList = leftToken.GetAncestor<TypeArgumentListSyntax>();
-                    if (typeArgumentList != null)
-                    {
-                        if (typeArgumentList.LessThanToken != default && typeArgumentList.GreaterThanToken != default)
-                        {
-                            testPosition = typeArgumentList.LessThanToken.SpanStart;
-                        }
-                    }
+                    testPosition = leftToken.GetAncestor<GenericNameSyntax>()?.SpanStart ?? testPosition;
                 }
+
+                var tokenLeftOfType = syntaxTree.FindTokenOnLeftOfPosition(testPosition, cancellationToken);
+                if (tokenLeftOfType.IsKind(SyntaxKind.DotToken) && tokenLeftOfType.Parent.IsKind(SyntaxKind.QualifiedName))
+                {
+                    testPosition = tokenLeftOfType.Parent.SpanStart;
+                    tokenLeftOfType = syntaxTree.FindTokenOnLeftOfPosition(testPosition, cancellationToken);
+                }
+                isRefType = TryMoveTestPositionIfRefType(tokenLeftOfType, ref testPosition);
+            }
+            else
+            {
+                isRefType = TryMoveTestPositionIfRefType(prevToken, ref testPosition);
             }
 
-            if ((!leftToken.GetPreviousTokenIfTouchingWord(position).IsKindOrHasMatchingText(SyntaxKind.AsyncKeyword) &&
-                syntaxTree.IsMemberDeclarationContext(testPosition, contextOpt: null, validModifiers: SyntaxKindSet.AllMemberModifiers, validTypeDeclarations: SyntaxKindSet.ClassInterfaceStructTypeDeclarations, canBePartial: false, cancellationToken: cancellationToken)) ||
+            var notAsync = isRefType || !prevToken.IsKindOrHasMatchingText(SyntaxKind.AsyncKeyword);
```

Comment:
i feel like needing to track `isRefTYpe` isn't really necessary for reasonable behavior of this feature.

Reply:
OK, I'll drop them.